queue_job/controllers/main.py

```diff
@@ -0,0 +1,132 @@
+        # or if something tries to run a job that is not enqueued
+        # before its execution, stop
+        if job.state != ENQUEUED:
+            _logger.warning('job %s is in state %s '
+                            'instead of enqueued in /runjob',
+                            job.uuid, job.state)
+            return
+
+        # TODO: set_started should be done atomically with
+        #       update queue_job set=state=started
+        #       where state=enqueid and id=
+        job.set_started()
+        job.store()
+        http.request.env.cr.commit()
+
+        _logger.debug('%s started', job)
+        job.perform()
+        job.set_done()
+        job.store()
+        http.request.env.cr.commit()
+        _logger.debug('%s done', job)
+
+    @http.route('/connector/runjob', type='http', auth='none')
+    def old_runjob(self, db, job_uuid, **kw):
+        _logger.warning('/connector/runjob is deprecated, the new route is'
+                        '/queue_job/runjob')
+        return self.runjob(db, job_uuid, **kw)
+
+    @http.route('/queue_job/runjob', type='http', auth='none')
+    def runjob(self, db, job_uuid, **kw):
+
+        env = http.request.env(user=odoo.SUPERUSER_ID)
+
+        def retry_postpone(job, message, seconds=None):
+            job.postpone(result=message, seconds=seconds)
+            job.set_pending(reset_retry=False)
+            job.store()
+            env.cr.commit()
+
+        job = self._load_job(env, job_uuid)
+        if job is None:
+            return ""
+        env.cr.commit()
+
+        try:
+            try:
+                self._try_perform_job(env, job)
+            except OperationalError as err:
+                # Automatically retry the typical transaction serialization
+                # errors
+                if err.pgcode not in PG_CONCURRENCY_ERRORS_TO_RETRY:
+                    raise
+
+                retry_postpone(job, tools.ustr(err.pgerror, errors='replace'),
+                               seconds=PG_RETRY)
+                _logger.debug('%s OperationalError, postponed', job)
+
+        except NothingToDoJob as err:
+            if unicode(err):
+                msg = unicode(err)
+            else:
+                msg = _('Job interrupted and set to Done: nothing to do.')
+            job.set_done(msg)
+            job.store()
+            env.cr.commit()
+
+        except RetryableJobError as err:
+            # delay the job later, requeue
+            retry_postpone(job, unicode(err), seconds=err.seconds)
+            _logger.debug('%s postponed', job)
+
+        except (FailedJobError, Exception):
+            buff = StringIO()
+            traceback.print_exc(file=buff)
+            _logger.error(buff.getvalue())
+            job.env.clear()
+            with odoo.api.Environment.manage():
+                with odoo.registry(job.env.cr.dbname).cursor() as new_cr:
+                    new_env = api.Environment(new_cr, job.env.uid,
```

Comment:
Do you mean we should replace `job.env.uid` by `SUPERUSER_ID`? I replaced by `job.env(cr=new_cr)` which returns the new env with only the `cr` modified, avoiding some repetition.